In [1]:
import pandas as pd
import sqlite3

# 9.1 Reading data from SQL databases

So far we've only talked about reading data from CSV files. That's a pretty common way to store data, but there are many others! Pandas can read from HTML, JSON, SQL, Excel (!!!), HDF5, Stata, and a few other things. In this chapter we'll talk about reading data from SQL databases.

You can read data from a SQL database using the `pd.read_sql` function. `read_sql` will automatically convert SQL column names to DataFrame column names.

`read_sql` takes 2 arguments: a `SELECT` statement, and a database connection object. This is great because it means you can read from *any* kind of SQL database -- it doesn't matter if it's MySQL, SQLite, PostgreSQL, or something else.

This example reads from a SQLite database, but any other database would work the same way.

In [3]:
con = sqlite3.connect("../data/weather_2012.sqlite")
df = pd.read_sql("SELECT temp as T, date_time as Time, id as ID from weather_2012 LIMIT 3", con)
df

T                 Time  ID
0 -1.8  2012-01-01 00:00:00   1
1 -1.8  2012-01-01 01:00:00   2
2 -1.8  2012-01-01 02:00:00   3

`read_sql` doesn't automatically set the primary key (`id`) to be the index of the dataframe. You can make it do that by adding an `index_col` argument to `read_sql`. 

If you've used `read_csv` a lot, you may have seen that it has an `index_col` argument as well. This one behaves the same way.

In [4]:
df = pd.read_sql("SELECT * from weather_2012 LIMIT 3", con, index_col='id')
df

date_time  temp
id                           
1   2012-01-01 00:00:00  -1.8
2   2012-01-01 01:00:00  -1.8
3   2012-01-01 02:00:00  -1.8

If you want your dataframe to be indexed by more than one column, you can give a list of columns to `index_col`:

In [5]:
df = pd.read_sql("SELECT * from weather_2012 LIMIT 3", con, 
                 index_col=['id', 'date_time'])
df

temp
id date_time                
1  2012-01-01 00:00:00  -1.8
2  2012-01-01 01:00:00  -1.8
3  2012-01-01 02:00:00  -1.8

# 9.2 Writing to a SQLite database

Pandas has a `write_frame` function which creates a database table from a dataframe. Right now this only works for SQLite databases. Let's use it to move our 2012 weather data into SQL.

You'll notice that this function is in `pd.io.sql`. There are a ton of useful functions for reading and writing various kind of data in `pd.io`, and it's worth spending some time exploring them. ([see the documentation!](http://pandas.pydata.org/pandas-docs/stable/io.html))

In [6]:
weather_df = pd.read_csv('../data/weather_2012.csv')
con = sqlite3.connect("../data/test_db.sqlite")
con.execute("DROP TABLE IF EXISTS weather_2012")
weather_df.to_sql("weather_2012", con)

C:\Users\Hamahmi\Anaconda3\lib\site-packages\pandas\core\generic.py:2127: UserWarning: The spaces in these column names will not be changed. In pandas versions < 0.14, spaces were converted to underscores.
  dtype=dtype)


We can now read from the `weather_2012` table in  `test_db.sqlite`, and we see that we get the same data back:

In [7]:
con = sqlite3.connect("../data/test_db.sqlite")
df = pd.read_sql("SELECT * from weather_2012 LIMIT 3", con)
df

index            Date/Time  Temp (C)  Dew Point Temp (C)  Rel Hum (%)  \
0      0  2012-01-01 00:00:00      -1.8                -3.9           86   
1      1  2012-01-01 01:00:00      -1.8                -3.7           87   
2      2  2012-01-01 02:00:00      -1.8                -3.4           89   

   Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)               Weather  
0                4              8.0           101.24                   Fog  
1                4              8.0           101.24                   Fog  
2                7              4.0           101.26  Freezing Drizzle,Fog

The nice thing about having your data in a database is that you can do arbitrary SQL queries. This is cool especially if you're more familiar with SQL. Here's an example of sorting by the Weather column:

In [10]:
con = sqlite3.connect("../data/test_db.sqlite")
df = pd.read_sql("SELECT * from weather_2012 GROUP BY Weather", con)
df

index            Date/Time  Temp (C)  Dew Point Temp (C)  Rel Hum (%)  \
0    8756  2012-12-30 20:00:00     -13.8               -16.5           80   
1    8766  2012-12-31 06:00:00      -9.7               -11.7           85   
2    8545  2012-12-22 01:00:00       1.2                 0.1           92   
3    8482  2012-12-19 10:00:00       0.6                 0.2           97   
4    8433  2012-12-17 09:00:00       0.4                -0.7           92   
5    8490  2012-12-19 18:00:00       1.2                 0.1           92   
6    8547  2012-12-22 03:00:00       1.1                 0.1           93   
7    8722  2012-12-29 10:00:00     -12.0               -13.3           90   
8    8424  2012-12-17 00:00:00      -2.3                -3.3           93   
9    8261  2012-12-10 05:00:00      -1.9                -2.8           94   
10    757  2012-02-01 13:00:00      -5.0                -7.7           81   
11   8690  2012-12-28 02:00:00      -4.3                -6.3           86   
12   1830  2012-03-17 06:00:00      -0.1                -0.4           98   
13   8426  2012-12-17 02:00:00      -2.2                -3.3           92   
14   8425  2012-12-17 01:00:00      -2.3                -3.3           93   
15    759  2012-02-01 15:00:00      -4.9                -7.5           82   
16   8427  2012-12-17 03:00:00      -2.6                -3.7           92   
17    297  2012-01-13 09:00:00      -5.0                -7.3           84   
18   8340  2012-12-13 12:00:00      -0.4                -2.9           83   
19   8758  2012-12-30 22:00:00     -13.7               -16.3           81   
20   8264  2012-12-10 08:00:00       1.7                 0.8           94   
21   8673  2012-12-27 09:00:00      -5.7                -6.7           93   
22   8676  2012-12-27 12:00:00      -5.5                -6.6           92   
23   8763  2012-12-31 03:00:00     -11.8               -14.4           81   
24   8541  2012-12-21 21:00:00       1.3                 0.4           94   
25   8363  2012-12-14 11:00:00       2.6                -2.0           72   
26   7035  2012-10-20 03:00:00      12.8                12.1           96   
27   8146  2012-12-05 10:00:00       2.1                -1.8           75   
28   8273  2012-12-10 17:00:00       0.1                -0.3           97   
29   1737  2012-03-13 09:00:00       4.4                 2.3           86   
30   8453  2012-12-18 05:00:00       0.6                -0.6           92   
31   8529  2012-12-21 09:00:00       0.8                -0.2           93   
32   8520  2012-12-21 00:00:00       1.9                -2.1           75   
33   8229  2012-12-08 21:00:00       0.8                 0.3           96   
34   8525  2012-12-21 05:00:00       1.3                 0.1           92   
35   8783  2012-12-31 23:00:00       0.0                -2.1           86   
36   7887  2012-11-24 15:00:00       0.7                -6.4           59   
37   8768  2012-12-31 08:00:00      -8.6               -10.3           87   
38   8725  2012-12-29 13:00:00     -10.0               -11.1           92   
39   8683  2012-12-27 19:00:00      -3.9                -5.6           88   
40   8770  2012-12-31 10:00:00      -7.4                -8.9           89   
41    765  2012-02-01 21:00:00      -4.3                -7.2           80   
42   8430  2012-12-17 06:00:00      -2.1                -3.2           92   
43   4729  2012-07-16 01:00:00      21.6                19.4           87   
44   3582  2012-05-29 06:00:00      10.9                 9.0           88   
45   4758  2012-07-17 06:00:00      19.6                18.5           93   
46   4914  2012-07-23 18:00:00      20.6                19.1           91   
47   6188  2012-09-14 20:00:00      16.5                15.8           96   
48   5108  2012-07-31 20:00:00      22.4                18.7           80   
49   4757  2012-07-17 05:00:00      20.6                18.6           88   

    Wind Spd (km/h)  Visibility (km)  Stn Press (kPa)  \
0                24

If you have a PostgreSQL database or MySQL database, reading from it works exactly the same way as reading from a SQLite database. You create a connection using `psycopg2.connect()` or `MySQLdb.connect()`, and then use

`pd.read_sql("SELECT whatever from your_table", con)`

# 9.3 Connecting to other kinds of database

To connect to a MySQL database:

*Note: For these to work, you will need a working MySQL / PostgreSQL database, with the correct localhost, database name, etc.*

To connect to a PostgreSQL database:

<style>
    @font-face {
        font-family: "Computer Modern";
        src: url('http://mirrors.ctan.org/fonts/cm-unicode/fonts/otf/cmunss.otf');
    }
    div.cell{
        width:800px;
        margin-left:16% !important;
        margin-right:auto;
    }
    h1 {
        font-family: Helvetica, serif;
    }
    h4{
        margin-top:12px;
        margin-bottom: 3px;
       }
    div.text_cell_render{
        font-family: Computer Modern, "Helvetica Neue", Arial, Helvetica, Geneva, sans-serif;
        line-height: 145%;
        font-size: 130%;
        width:800px;
        margin-left:auto;
        margin-right:auto;
    }
    .CodeMirror{
            font-family: "Source Code Pro", source-code-pro,Consolas, monospace;
    }
    .text_cell_render h5 {
        font-weight: 300;
        font-size: 22pt;
        color: #4057A1;
        font-style: italic;
        margin-bottom: .5em;
        margin-top: 0.5em;
        display: block;
    }
    
    .warning{
        color: rgb( 240, 20, 20 )
        }  